types/rdf-validate-shacl/index.d.ts

```diff
@@ -3,23 +3,25 @@
 // Definitions by: Tomasz Pluskiewicz <https://github.com/tpluscode>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
-import * as RDF from 'rdf-js';
-import DataFactory = require('./src/data-factory');
+import { DataFactory, DatasetCore, DatasetCoreFactory } from 'rdf-js';
 import ValidationReport = require('./src/validation-report');
 
+type FactoryFor<D> = D extends DatasetCore<infer OutQuad, infer InQuad> ? DataFactory<OutQuad, InQuad> & DatasetCoreFactory<OutQuad, InQuad, D> : never;
+type OutQuadOf<T> = T extends DatasetCore<infer Q, any> ? Q : never;
+
 declare namespace SHACLValidator {
-    interface Options {
-        factory?: RDF.DataFactory & RDF.DatasetCoreFactory | undefined;
+    interface Options<F extends DataFactory & DatasetCoreFactory> {
+        factory?: F | undefined;
         maxErrors?: number | undefined;
     }
 }
 
-declare class SHACLValidator {
-    constructor(shapes: RDF.DatasetCore, options?: SHACLValidator.Options);
-    factory: DataFactory;
+declare class SHACLValidator<D extends DatasetCore, F extends FactoryFor<D>> {
```

Comment:
Here too please add default types so that the interface can be used without type arguments

Also, it appears to be irrelevant whether the factory and dataset are related. I think the `FactoryFor<>` redundant

```suggestion
declare class SHACLValidator<D extends DatasetCore = DatasetCore, F extends DataFactory & DatasetCoreFactory = DataFactory & DatasetCoreFactory> {
```